### PR TITLE
[Core/AreaTriggers] Fix area trigger enter, exit and rest flags

### PR DESCRIPTION
**Pull request by jasongdove**
This implements the new `entered` area trigger flag, so expected actions (rested, quest push) are only taken when an area trigger is entered.

The rest system has also been updated to use flags like TCPP/TC, and to simplify calls to `Player::SetRestFlag`.

Steps to verify tavern rest:

1. Create horde character
2. `.debug areatriggers`
3. `.tele tranquillien`
4. Walk into building
5. See area trigger enter message
6. See rested state
7. Leave building
8. See area trigger left (exited) message
9. See rested state has stopped

Steps to verify quest push:

1. Create horde character
2. `.debug areatriggers`
3. `.levelup 79`
4. `.tele org`
5. See area trigger enter message
6. See quest push for `A Personal Summons`

### CODE REVIEW

File: src/server/game/DataStores/DBCStructure.h [hunk 1]
```diff
@@ -621,22 +621,20 @@ struct AreaPOIEntry
 
 struct AreaTriggerEntry
```
Comment (jasongdove) [inline src/server/game/DataStores/DBCStructure.h:622]: See https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/blob/7316f67b283f4dfc6625275d5b4e96c33b1051cd/src/server/game/DataStores/DBCStructure.h#L255-L268 and 6.x https://github.com/TrinityCore/TrinityCore/blob/8fa52860ab666c95579f6c44bca26a49f7414b14/src/server/game/DataStores/DBCStructure.h#L87-L89

File: src/server/game/Entities/Player/Player.h [hunk 2]
```diff
@@ -2768,14 +2737,13 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
     {
         return GetRestTime() >= 10 * IN_MILLISECONDS;
     }
-    uint32 GetXPRestBonus(uint32 xp);
     uint32 GetRestTime() const
     {
```
Comment (jasongdove) [inline src/server/game/Entities/Player/Player.h:2741]: `GetRestTime` and `SetRestTime` are only used by Lua and had their own `m_restTime` which wasn't actually used by the rested system. Fixed to use `_restTime`.

File: src/server/game/Handlers/MiscHandler.cpp [hunk 3]
```diff
@@ -1119,152 +1119,117 @@ void WorldSession::HandleResurrectResponseOpcode(WorldPacket& recvData)
...
+    if (sObjectMgr->IsTavernAreaTrigger(areaTriggerId))
     {
         // set resting flag we are in the inn
-        player->SetFlag(PLAYER_FIELD_PLAYER_FLAGS, PLAYER_FLAGS_RESTING);
-        player->InnEnter(time(NULL), atEntry->mapid, atEntry->x, atEntry->y, atEntry->z);
-        player->SetRestType(REST_TYPE_IN_TAVERN);
+        if (entered)
+            player->SetRestFlag(REST_FLAG_IN_TAVERN, atEntry->ID);
+        else
+            player->RemoveRestFlag(REST_FLAG_IN_TAVERN);
```
Comment (jasongdove) [inline src/server/game/Handlers/MiscHandler.cpp:1211]: This fixes the bug with rested state after tavern exit.

File: src/server/game/Miscellaneous/Language.h [hunk 4]
```diff
@@ -1146,6 +1146,7 @@ enum TrinityStrings
     LANG_BG_TOK_PICKEDUP_ORANGE          = 1356,
 
     // FREE IDS                           1334-1999
+    LANG_DEBUG_AREATRIGGER_LEFT          = 1999,
```
Comment (jasongdove) [inline src/server/game/Miscellaneous/Language.h:1149]: TC uses this ID for this message.

File: src/server/game/OutdoorPvP/OutdoorPvP.h [hunk 5]
```diff
@@ -208,7 +208,7 @@ class OutdoorPvP : public ZoneScript
         virtual void FillInitialWorldStates(WorldStateBuilder& /*builder*/) { }
 
         // called when a player triggers an areatrigger
-        virtual bool HandleAreaTrigger(Player* player, uint32 trigger);
+        virtual bool HandleAreaTrigger(Player* player, uint32 trigger, bool entered);
```
Comment (leelf00) [inline src/server/game/OutdoorPvP/OutdoorPvP.h:211]: What is param `entered` usage?

File: src/server/game/Battlegrounds/Battleground.h [hunk 6]
```diff
@@ -518,7 +518,7 @@ class Battleground
 
         // Triggers handle
         // must be implemented in BG subclass
-        virtual void HandleAreaTrigger(Player* /*player*/, uint32 /*Trigger*/);
+        virtual void HandleAreaTrigger(Player* /*player*/, uint32 /*Trigger*/, bool /*Entered*/);
```
Comment (leelf00) [inline src/server/game/Battlegrounds/Battleground.h:521]: What is param `Entered` usage?

File: src/server/game/OutdoorPvP/OutdoorPvP.h (hunk 5 above)
Comment (jasongdove) [reply src/server/game/OutdoorPvP/OutdoorPvP.h:211]: When the client sends area trigger message, `entered` field is `true` when going from outside the area to inside the area, and `entered` field is `false` when going from inside the area to outside the area. 

For a tavern/inn, entered = true (walked into inn) means set rested flag, entered = false (walked outside of inn) means remove rested flag.

The field isn't used in this PR in these other area trigger calls, but I think it should be available.

Does that answer your question?

Comment (leelf00) [reply src/server/game/OutdoorPvP/OutdoorPvP.h:211]: This is correct. I will test the whole PR tonight.

**Pull request merged**